plan.go

```diff
@@ -19,26 +20,60 @@ type QueryPlanStep struct {
 	SelectionSet   ast.SelectionSet
 	InsertionPoint []string
 	Then           []*QueryPlanStep
+
+	executionResult *executionStepResult
+}
+
+type executionStepResult struct {
+	executed  bool
+	error     error
+	timeTaken time.Duration
+}
+
+func (e *executionStepResult) MarshalJSON() ([]byte, error) {
+	type toMarshal struct {
+		Executed  bool
+		Error     error `json:",omitempty"`
+		TimeTaken string
+	}
+	j := toMarshal{
+		Executed:  e.executed,
+		TimeTaken: e.timeTaken.String(),
+	}
+
+	if e.error != nil {
+		j.Error = e.error
+	}
+
+	return json.Marshal(&j)
 }
 
 // MarshalJSON marshals the step the JSON
 func (s *QueryPlanStep) MarshalJSON() ([]byte, error) {
 	ctx := graphql.WithOperationContext(context.Background(), &graphql.OperationContext{
 		Variables: map[string]interface{}{},
 	})
-	return json.Marshal(&struct {
-		ServiceURL     string
-		ParentType     string
-		SelectionSet   string
-		InsertionPoint []string
-		Then           []*QueryPlanStep
-	}{
+	type toMarshal struct {
```

Comment:
Is it worth extracting these out to locally named types instead of the anonymous structs?
The only conditional assignment is based on if they are nil values so the `omitempty` tag on the field will handle that.

Reply:
They are named types, just scoped to the function. I don't see a need to move them out if they're only used for json marshalling. As to why they're there, that pattern was already used and I do like keeping all of that information unexported and only exposing it as we marshal.

Reply:
Oh I didn't mean to extract them, I meant is there any benefit to giving them local names inside the function when they were anonymous before. i.e go back to
```golang
return json.Marshal(&struct{
	ServiceURL          string
	ParentType          string
	SelectionSet        string
	InsertionPoint      []string
	ExecutionStepResult *executionStepResult `json:",omitempty"`
	Then                []*QueryPlanStep
}{
	ServiceURL:          s.ServiceURL,
	ParentType:          s.ParentType,
	SelectionSet:        formatSelectionSetSingleLine(ctx, nil, s.SelectionSet),
	InsertionPoint:      s.InsertionPoint,
	Then:                s.Then,
	ExecutionStepResult: s.executionResult,
})
```

Reply:
Yeah right, no reason. I've changed it back to anonymous structs.